flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1697,133 +1621,620 @@ class ExprLowering {
+          builder.restoreInsertionPoint(insPt);
+          return iterSpace.outerResult();
+        },
+        exp.u);
+  }
+
+  std::pair<IterationSpace, mlir::OpBuilder::InsertPoint> genIterSpace() {
+    auto loc = getLoc();
+    auto idxTy = builder.getIndexType();
+    auto zero = builder.createIntegerConstant(loc, idxTy, 0);
+    auto one = builder.createIntegerConstant(loc, idxTy, 1);
+    llvm::SmallVector<mlir::Value, 8> loopUppers;
+
+    // Convert the shape to closed interval form.
+    for (auto s : destination.getExtents()) {
+      auto ub = builder.createConvert(loc, idxTy, s);
+      auto up = builder.create<mlir::SubIOp>(loc, ub, one);
+      loopUppers.push_back(up);
+    }
+
+    // Iteration space is created with outermost columns, innermost rows
+    fir::DoLoopOp inner;
+    auto innerArg = destination.getResult();
+    mlir::Value outerRes;
+    llvm::SmallVector<mlir::Value, 8> ivars;
+    auto insPt = builder.saveInsertionPoint();
+    const auto loopDepth = loopUppers.size();
+    assert(loopDepth > 0);
+    for (auto i : llvm::enumerate(llvm::reverse(loopUppers))) {
+      if (i.index() > 0) {
+        assert(inner);
+        builder.setInsertionPointToStart(inner.getBody());
+      }
+      auto loop = builder.create<fir::DoLoopOp>(
+          loc, zero, i.value(), one, /*unordered=*/false,
+          /*finalCount=*/false, mlir::ValueRange{innerArg});
+      innerArg = loop.getRegionIterArgs().front();
+      ivars.push_back(loop.getInductionVar());
+      if (!outerRes)
+        outerRes = loop.getResult(0);
+      if (!inner)
+        insPt = builder.saveInsertionPoint();
+      if (i.index() < loopDepth - 1) {
+        // Add the fir.result for all loops except the innermost one.
+        builder.setInsertionPointToStart(loop.getBody());
+        builder.create<fir::ResultOp>(loc, innerArg);
+      }
+      inner = loop;
+    }
+
+    // move insertion point inside loop nest
+    builder.setInsertionPointToStart(inner.getBody());
+    return {IterationSpace{innerArg, outerRes, ivars}, insPt};
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Expression traversal and lowering.
+  //===--------------------------------------------------------------------===//
+
+  // Lower the expression in a scalar context.
+  template <typename A>
+  ExtValue asScalar(const A &x) {
+    return ScalarExprLowering{getLoc(), converter, symMap, stmtCtx}.genval(x);
   }
 
+  // An expression with non-zero rank is an array expression.
   template <typename A>
-  void prelude(const Fortran::evaluate::Expr<A> &exp) {
-    std::visit([&](const auto &e) { genpre(e); }, exp.u);
+  static bool isArray(const A &x) {
+    return x.Rank() != 0;
   }
 
   template <typename A, typename = std::enable_if_t<Fortran::common::HasMember<
                             A, Fortran::evaluate::TypelessExpression>>>
-  void genpre(const A &) {}
+  CC genarr(const A &x) {
```

Comment:
It is still unclear to me which case falls here since `genarr(const Fortran::evaluate::Expr<A> &x)` should be dispacthing between non-scalar and scalars. It seems to me that if we fall here, that is most likely a TODO and not a scalar expression. I am OK with leaving this in if it helps scaffolding, but if this is indeed scaffolding, it may deserve a comment so that we do not forget to remove it/make it an error.
[edit: I think derived type array expressions are the ones that would currently fall here]

Reply:
Is it not legal to have `A + z'deadbeef'` where A has rank > 0?